SSA/Projects/LLVMAndRiscv.lean

```
@@ -125,4 +127,132 @@ attribute [simp, simp_denote] outTy_map_signature_eq
+              | .riscv originalRISCVTy =>
+                if hty' : originalRISCVTy = t then
+                  return ⟨v.val, by rw [h, hty']⟩
+                else
+                  throw <|.generic s!"INTERNAL ERROR: This case is impossible, RISCV expression is pointing to an incorrect bitwidth LLVM argument."
+            | none =>
+              -- this is impossible, because ctxtTransformToLLVM is a `List.map`, which always maintains length.
+              throw <| .generic s!"INTERNAL ERROR: This case is impossible, as 'ctxtTransformToLLVM' is length-preserving."
+        return Expr.mk
+          (op := Op.riscv op1)
+          (eff_le := eff_le1)
+          (ty_eq := ty_eq1 ▸ rfl)
+          (args := args')
+          (regArgs := HVector.nil)
+
+instance : MLIR.AST.TransformTy LLVMPlusRiscV 0 where
+  mkTy tStx := do
+    try
+      let llvmParse ← InstcombineTransformDialect.mkTy tStx
+      return .llvm llvmParse
+    catch _llvmErr =>
+      try
+        let riscvParse ← RiscvMkExpr.mkTy tStx
+        return .riscv riscvParse
+      catch _riscvErr =>
+        throw <| .generic s!"Unable to parse type as either LLVM type or RISCV type."
+
+def mkExpr (Γ : Ctxt _) (opStx : MLIR.AST.Op 0) :
+  MLIR.AST.ReaderM (LLVMPlusRiscV) (Σ eff ty, Expr LLVMPlusRiscV Γ eff ty) := do
+  if (opStx.name = "builtin.unrealized_conversion_cast") then
+    match opStx.args with
+    | v₁Stx :: [] =>
+      let ⟨ty₁, v₁⟩ ← MLIR.AST.TypedSSAVal.mkVal Γ v₁Stx
+      match ty₁ with
+      | .riscv (.bv) =>
+              return ⟨ .pure,  .llvm (.bitvec 64) ,⟨ .castRiscv , by rfl, by constructor,
+               .cons v₁ <| .nil,
+                .nil⟩⟩
+      | .llvm (.bitvec 64) =>
+              return ⟨ .pure, .riscv (.bv) ,⟨ .castLLVM , by rfl, by constructor,
+               .cons v₁ <| .nil,
+                .nil⟩⟩
+      | _ => throw <| .unsupportedOp s!"unsupported operation {repr opStx}"
+    | _ => throw <| .unsupportedOp s!"unsupported operation {repr opStx}"
```

Note that unsupportedOp is not the right error here: this would claim that `builtin.unrealized_conversion_cast` is an unsupported operation, even though it most definitely is supported. I believe the actual error is that the number of arguments (i.e, the arity) is wrong, right? So throw an error that says that! (Although, if you follow my last comment and use `assumeArity`, you won't even have to throw the error yourself, the helper should throw the right error for you!)

---

changed the error to be an arity error